File: app/src/main/java/wxm/KeepAccount/db/UsrDBUtility.java
package wxm.KeepAccount.db;

import com.j256.ormlite.dao.RuntimeExceptionDao;

import org.greenrobot.eventbus.EventBus;

import java.util.List;

import cn.wxm.andriodutillib.DBHelper.DBUtilityBase;
import cn.wxm.andriodutillib.util.MD5Util;
import cn.wxm.andriodutillib.util.UtilFun;
import wxm.KeepAccount.define.GlobalDef;
import wxm.KeepAccount.define.UsrItem;
import wxm.KeepAccount.ui.utility.NoteDataHelper;
import wxm.KeepAccount.utility.ContextUtil;

/**
 * usr数据处理类
 * Created by 123 on 2016/8/9.
 */
public class UsrDBUtility extends DBUtilityBase<UsrItem, Integer> {
    private final String TAG = "UsrDBUtility";

    public UsrDBUtility() {
        super();
    }

    @Override
    protected RuntimeExceptionDao<UsrItem, Integer> getDBHelper() {
        return ContextUtil.getDBHelper().getUsrItemREDao();
    }

    /**
     * 检查是否用户'usr'已经存在
     *
     * @param usr 待检查用户
     * @return 如果存在返回true, 否则返回false
     */
    public boolean hasUsr(String usr) {
        if (UtilFun.StringIsNullOrEmpty(usr))
            return false;

        List<UsrItem> ret = getDBHelper().queryForEq(UsrItem.FIELD_NAME, usr);
        return !((null == ret) || (ret.size() < 1));
    }


    /**
     * 添加用户
     *
     * @param usr 待添加用户名
     * @param pwd 待添加用户密码
     * @return 添加成功返回添加后的数据, 否则返回null
     */
    public UsrItem addUsr(String usr, String pwd) {
        if (UtilFun.StringIsNullOrEmpty(usr) || UtilFun.StringIsNullOrEmpty(pwd))
            return null;

        String pwdpad = pwd;
        if (pwdpad.length() < GlobalDef.STR_PWD_PAD.length()) {
            pwdpad += GlobalDef.STR_PWD_PAD.substring(pwd.length());
        }
        pwdpad = MD5Util.string2MD5(pwdpad);

        UsrItem uiret = null;
        List<UsrItem> ret = getDBHelper().queryForEq(UsrItem.FIELD_NAME, usr);
        if ((null == ret) || (ret.size() < 1)) {
            UsrItem ui = new UsrItem();
            ui.setName(usr);
            ui.setPwd(pwdpad);

            if (createData(ui))
                uiret = ui;
        } else {
            UsrItem uiold = ret.get(0);
            if (!pwdpad.equals(uiold.getPwd())) {
                uiold.setPwd(pwdpad);
                if (modifyData(uiold))
                    uiret = uiold;
            } else {
                uiret = uiold;
            }
        }

        return uiret;
    }


    /**
     * 检查用户
     *
     * @param usr 待检查用户名
     * @param pwd 待检查用户密码
     * @return 如果符合返回true, 否则返回false
     */
    public boolean checkUsr(String usr, String pwd) {
        String pwdpad = pwd;
        if (pwdpad.length() < GlobalDef.STR_PWD_PAD.length()) {
            pwdpad += GlobalDef.STR_PWD_PAD.substring(pwd.length());
        }

        List<UsrItem> lsui = getDBHelper().queryForEq(UsrItem.FIELD_NAME, usr);
        if ((null == lsui) || (lsui.size() < 1))
            return false;

        String checkPwd = MD5Util.string2MD5(pwdpad);
        return checkPwd.equals(lsui.get(0).getPwd());

    }


    /**
     * 检查登录信息，如果有符合的记录就返回对应用户信息
     *
     * @param usr 待检查用户名
     * @param pwd 待检查用户密码
     * @return 如果符合返回注册用户数据, 否则返回null
     */
    public UsrItem CheckAndGetUsr(String usr, String pwd) {
        String pwdpad = pwd;
        if (pwdpad.length() < GlobalDef.STR_PWD_PAD.length()) {
            pwdpad += GlobalDef.STR_PWD_PAD.substring(pwd.length());
        }

        List<UsrItem> lsui = getDBHelper().queryForEq(UsrItem.FIELD_NAME, usr);
        if ((null == lsui) || (lsui.size() < 1))
            return null;

        String checkPwd = MD5Util.string2MD5(pwdpad);
        if (checkPwd.equals(lsui.get(0).getPwd()))
            return lsui.get(0);

        return null;
    }

    /**
     * 使用用户信息登录APP
     * （并刷新数据)
     *
     * @param usr 待检查用户名
     * @param pwd 待检查用户密码
     * @return 如果登录成功返回true, 否则返回false
     */
    public boolean loginByUsr(String usr, String pwd) {
        UsrItem ui = CheckAndGetUsr(usr, pwd);
        if (null == ui)
            return false;

        ContextUtil.setCurUsr(ui);
        NoteDataHelper.getInstance().refreshData();
        return true;
    }

    @Override
    protected void onDataModify(List<Integer> md) {
        EventBus.getDefault().post(new DBDataChangeEvent());
    }

    @Override
    protected void onDataCreate(List<Integer> cd) {
        EventBus.getDefault().post(new DBDataChangeEvent());
    }

    @Override
    protected void onDataRemove(List<Integer> dd) {
        EventBus.getDefault().post(new DBDataChangeEvent());
    }
}

File: app/src/main/java/wxm/KeepAccount/db/UsrDBUtility.kt
package wxm.KeepAccount.db

import com.j256.ormlite.dao.RuntimeExceptionDao
import org.greenrobot.eventbus.EventBus
import wxm.KeepAccount.define.GlobalDef
import wxm.KeepAccount.item.UsrItem
import wxm.KeepAccount.utility.AppUtil
import wxm.androidutil.db.DBUtilityBase
import wxm.androidutil.util.MD5Util
import wxm.androidutil.improve.doJudge
import wxm.androidutil.improve.forObj

/**
 * usr数据处理类
 * Created by WangXM on 2016/8/9.
 */
class UsrDBUtility : DBUtilityBase<UsrItem, Int>() {
    override fun onDataModify(md: List<Int>) {
        EventBus.getDefault().post(DBDataChangeEvent())
    }

    override fun onDataCreate(cd: List<Int>) {
        EventBus.getDefault().post(DBDataChangeEvent())
    }

    override fun onDataRemove(dd: List<Int>) {
        EventBus.getDefault().post(DBDataChangeEvent())
    }

    override fun getDBHelper(): RuntimeExceptionDao<UsrItem, Int> {
        return AppUtil.dbHelper.usrItemREDao
    }

    /**
     * 检查是否用户'usr'已经存在
     * @param usr 待检查用户
     * @return 如果存在返回true, 否则返回false
     */
    fun hasUsr(usr: String): Boolean {
        return dbHelper.queryBuilder().setCountOf(true)
                .where().eq(UsrItem.FIELD_NAME, usr).prepare().let {
                    dbHelper.countOf(it)
                } >= 1
    }


    /**
     * 添加用户
     * @param usr 待添加用户名
     * @param pwd 待添加用户密码
     * @return 添加成功返回添加后的数据, 否则返回null
     */
    fun addUsr(usr: String, pwd: String): UsrItem? {
        if (usr.isEmpty() || pwd.isEmpty())
            return null

        if (hasUsr(usr))
            return null

        return UsrItem().apply {
            this.name = usr
            this.pwd = getMd5Pwd(pwd)
        }.let {
            createData(it).doJudge(it, null)
        }
    }


    /**
     * 检查用户
     * @param usr 待检查用户名
     * @param pwd 待检查用户密码
     * @return 如果符合返回true, 否则返回false
     */
    fun checkUsr(usr: String, pwd: String): Boolean {
        val query = dbHelper.queryBuilder().where()
                .eq(UsrItem.FIELD_NAME, usr)
                .and()
                .eq(UsrItem.FIELD_PWD, getMd5Pwd(pwd))
                .prepare()

        return dbHelper.query(query).let {
            null != it && it.isNotEmpty()
        }
    }


    /**
     * 检查登录信息，如果有符合的记录就返回对应用户信息
     * @param usr 待检查用户名
     * @param pwd 待检查用户密码
     * @return 如果符合返回注册用户数据, 否则返回null
     */
    fun checkGetUsr(usr: String, pwd: String): UsrItem? {
        val query = dbHelper.queryBuilder().where()
                .eq(UsrItem.FIELD_NAME, usr)
                .and()
                .eq(UsrItem.FIELD_PWD, getMd5Pwd(pwd))
                .prepare()

        return dbHelper.query(query).let {
            if (null == it || it.isEmpty()) null
            else it[0]
        }
    }

    /**
     * use usr name [usr] and password [pwd] login app
     * and record in history
     *
     * return true if everything ok
     */
    fun loginByUsr(usr: String, pwd: String): Boolean {
        return checkGetUsr(usr, pwd).forObj(
                { loginByUsr(it, true) }, { false })
    }

    /**
     * use usr [usr] login app
     * if [recordHistory] is true, record login history
     *
     * return true if everything ok
     */
    fun loginByUsr(usr: UsrItem, recordHistory: Boolean): Boolean {
        AppUtil.curUsr = usr
        if (recordHistory) LoginHistoryUtility.addHistory(usr)
        return true
    }

    /**
     * change [usr] icon to [fn]
     *
     * return true if everything ok
     */
    fun changeIcon(usr: UsrItem, fn: String): Boolean {
        return dbHelper.updateBuilder().let {
            it.updateColumnValue(UsrItem.FIELD_ICON_PATH, fn)
            it.where().eq(UsrItem.FIELD_ID, usr.id)
            it.update() == 1
        }.doJudge(
                {
                    usr.iconPath = AppUtil.usrUtility.getData(usr.id)!!.iconPath
                    true
                },
                { false }
        )
    }

    /**
     * return true if [pwd] can use as pwd
     */
    fun pwdValidity(pwd: String): Int {
        return (pwd.length > 4).doJudge(RET_OK, RET_PWD_TO_SHORT)
    }

    fun changePwd(usr: UsrItem, pwd: String): Boolean {
        if (RET_OK != pwdValidity(pwd)) {
            return false
        }

        return dbHelper.updateBuilder().let {
            it.updateColumnValue(UsrItem.FIELD_PWD, getMd5Pwd(pwd))
                    .where().eq(UsrItem.FIELD_ID, usr.id)

            it.update() == 1
        }.doJudge(
                {
                    usr.pwd = AppUtil.usrUtility.getData(usr.id)!!.pwd
                    true
                },
                { false }
        )
    }

    /**
     * return [orgPwd] as MD5 string
     */
    private fun getMd5Pwd(orgPwd: String): String {
        return (orgPwd.length < GlobalDef.STR_PWD_PAD.length).doJudge(
                { orgPwd + GlobalDef.STR_PWD_PAD.substring(orgPwd.length) },
                { orgPwd }
        ).let {
            MD5Util.string2MD5(it)
        }
    }

    companion object {
        const val RET_OK = 0
        const val RET_PWD_TO_SHORT = 1
    }
}
